src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceFeatures.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.neo3d;
+
+import org.jetbrains.annotations.ApiStatus;
+
+public interface GpuDeviceFeatures {
+    /**
+     * LogicOp is unusably problematic on Qualcomm GPUs via OpenGL
+     * LogicOp is unavailable on MacOS via Vulkan
+     */
+    boolean logicOp();
+
+    record Immutable(boolean logicOp) implements GpuDeviceFeatures {
+        /**
+         * Immutable record should only be constructed from an instance of the interface
+         * <br>
+         * Additional elements may be added without being considered a breaking change
+         *
+         * @see Immutable#Immutable(GpuDeviceFeatures)
+         */
+        @ApiStatus.Internal
+        public Immutable {}
+
+        public Immutable(GpuDeviceFeatures features) {
+            this(features.logicOp());
+        }
+    }
```

Comment:
Why is this in the API? It's trivial for each backend to replicate.

Reply:
currently its trivial, it wont stay that way. the one for properties is already not _completely_ trivial.

i haven't done this yet, but the calls in GpuDevice should be changed to return these records (currently returning the interfaces) so that in most cases the vtable lookup can be skipped.

Reply:
yea, just making the record package-private is probably the better option, where backends have to implement their own.

Reply:
records moved and made package-private